src/mongoc/mongoc-topology-description.c

```diff
@@ -0,0 +1,996 @@
+ *
+ * Side effects:
+ *       None.
+ *
+ *--------------------------------------------------------------------------
+ */
+
+mongoc_server_description_t *
+mongoc_topology_description_server_by_id (mongoc_topology_description_t *description,
+                                           uint32_t                       id)
+{
+   return mongoc_set_get(description->servers, id);
+}
+
+/*
+ *--------------------------------------------------------------------------
+ *
+ * _mongoc_topology_description_remove_server --
+ *
+ *       If present, remove this server from this topology description.
+ *
+ * Returns:
+ *       None.
+ *
+ * Side effects:
+ *       Removes server from topology's list of servers.
+ *
+ *--------------------------------------------------------------------------
+ */
+static void
+_mongoc_topology_description_remove_server (mongoc_topology_description_t *description,
+                                            mongoc_server_description_t   *server)
+{
+   if (description->cb.rm) {
+      description->cb.rm(server);
+   }
+
+   mongoc_set_rm(description->servers, server->id);
+}
+
+/*
+ *--------------------------------------------------------------------------
+ *
+ * _mongoc_topology_description_topology_has_server --
+ *
+ *       Return a server id if @server is in the given topology,
+ *       0 otherwise.
+ *
+ * Returns:
+ *       A server id, or 0.
+ *
+ * Side effects:
+ *       None.
+ *
+ *--------------------------------------------------------------------------
+ */
+static uint32_t
+_mongoc_topology_description_has_server (mongoc_topology_description_t *description,
+                                         const char                    *address)
+{
+   mongoc_server_description_t *server_iter;
+   mongoc_set_t *set = description->servers;
+   int i;
+
+   for (i = 0; i < set->items_len; i++) {
+      server_iter = set->items[i].item;
+      if (address == server_iter->connection_address) {
+         return server_iter->id;
+      }
+   }
+
+   return 0;
+}
+
+/*
+ *--------------------------------------------------------------------------
+ *
+ * _mongoc_topology_description_label_unknown_member --
+ *
+ *       Find the server description with the given @address and if its
+ *       type is UNKNOWN, set its type to @type.
+ *
+ * Returns:
+ *       None.
+ *
+ * Side effects:
+ *       None.
+ *
+ *--------------------------------------------------------------------------
+ */
+static void
+_mongoc_topology_description_label_unknown_member (mongoc_topology_description_t *description,
+                                                   const char *address,
+                                                   mongoc_server_description_type_t type)
+{
+   mongoc_server_description_t *server_iter;
+   mongoc_set_t *set = description->servers;
+   int i;
+
+   for (i = 0; i < set->items_len; i++) {
+      server_iter = set->items[i].item;
+      if (server_iter->connection_address == address &&
```

Comment:
strcmp, not "==", and normalize case according to spec:

https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst#hostnames-are-normalized-to-lower-case